### PR TITLE
2/n inter batch parallelism

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds DataFetcher to `data_connector` and loops. 

Parts #8867
Parts #8316

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -234,3 +233,8 @@ def should_accumulate(self) -> bool:
 
     def teardown(self) -> None:
         self.epoch_loop.teardown()
+
+    def _prepare_dataloader_iter(self) -> Iterator:
+        dataloader = self.trainer.accelerator.process_dataloader(self.trainer.train_dataloader)
+        dataloader = self.trainer.data_connector.get_profiled_dataloader(dataloader)
+        return iter(dataloader)
```
Comment (justusschock) [inline pytorch_lightning/loops/fit_loop.py:240]: can we have this as a utility function outside? Currently this is duplicated in evaluation_loop

File: pytorch_lightning/loops/processors/iterator_batch_processor.py [hunk 2]
```diff
@@ -130,6 +130,7 @@ def run(self, dataloader_iter: Iterator) -> Optional[AttributeDict]:
 
         with self.trainer.profiler.profile("model_forward"):
             with self.trainer.profiler.profile("training_step"):
+                step_kwargs = self._build_kwargs(dataloader_iter, batch_idx)
```
Comment (justusschock) [inline pytorch_lightning/loops/processors/iterator_batch_processor.py:133]: this is duplicated in line 129
```suggestion
```

File: pytorch_lightning/loops/processors/iterator_batch_processor.py [hunk 3]
```diff
@@ -158,16 +161,18 @@ def _build_kwargs(self, dataloader_iter: Iterator, batch_idx: int) -> Dict[str,
         """Builds the keyword arguments for training_step
 
         Args:
-            dataloader_iter: The dataloader to pass
+            batch: the batch to train on
```
Comment (justusschock) [inline pytorch_lightning/loops/processors/iterator_batch_processor.py:164]: This doesn't seem to be correct. It still gets the dataloader_iter (at least in the call above)

File: pytorch_lightning/loops/processors/iterator_batch_processor.py [hunk 4]
```diff
@@ -145,6 +146,8 @@ def run(self, dataloader_iter: Iterator) -> Optional[AttributeDict]:
         if training_step_output:
             batch_outputs[0].append(training_step_output)
 
+        if training_step_output:
+            batch_outputs[0].append(training_step_output)
```
Comment (carmocca) [inline pytorch_lightning/loops/processors/iterator_batch_processor.py:150]: Bad merge ;)

```suggestion
```

File: pytorch_lightning/loops/processors/iterator_batch_processor.py (hunk 2 above)
Comment (carmocca) [inline pytorch_lightning/loops/processors/iterator_batch_processor.py:133]: Bad merge

```suggestion
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 5]
```diff
@@ -66,15 +79,38 @@ def on_trainer_init(
...
-    def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
-        # FIXME: Temporary hack
-        if isinstance(self.data_fetcher, InterBatchParallelDataFetcher):
-            self.data_fetcher.setup(train_dataloader, batch_to_device=self.trainer.accelerator.batch_to_device)
+    def _check_training_step_requires_dataloader_iter(self) -> bool:
+        if not self.trainer.training:
+            return False
+        training_step_fx = getattr(self.trainer.lightning_module, "training_step")
+        contains_dataloader_iter = is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True)
+        return contains_dataloader_iter
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/data_connector.py:87]: This is not required for inter batch parallelism, right? This is just another flavor of training step.

File: pytorch_lightning/loops/processors/iterator_batch_processor.py [hunk 6]
```diff
@@ -158,16 +161,18 @@ def _build_kwargs(self, dataloader_iter: Iterator, batch_idx: int) -> Dict[str,
...
+            the keyword arguments for the training step
         """
         # enable not needing to add opt_idx to training_step
-        step_kwargs = OrderedDict([("dataloader_iter", dataloader_iter)])
+        step_kwargs = OrderedDict({"dataloader_iter": dataloader_iter})
+
+        lightning_module = self.trainer.lightning_module
 
-        training_step_fx = getattr(self.trainer.lightning_module, "training_step")
+        training_step_fx = getattr(lightning_module, "training_step")
```
Comment (carmocca) [inline pytorch_lightning/loops/processors/iterator_batch_processor.py:175]: Bad merge

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 7]
```diff
@@ -66,15 +79,38 @@ def on_trainer_init(
...
+
+    def _select_data_fetcher(self) -> AbstractDataFetcher:
+        if self._check_training_step_requires_dataloader_iter():
+            rank_zero_warn(
+                "Found `dataloader_iter` argument in the `training_step`. Note that the support for "
+                "this signature is experimental and the behavior may subject to change."
+            )
+            return DataLoaderIterDataFetcher()
+        elif self.trainer.training_type_plugin.on_gpu and os.getenv("PL_INTER_BATCH_PARALLELISM", "0") == "1":
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/data_connector.py:96]: I hope there will be a better way to enable this than the environment variable in the future?

File: tests/utilities/test_fetching.py [hunk 8]
```diff
@@ -27,7 +29,7 @@
 from tests.helpers.runif import RunIf
 
 
-@pytest.mark.parametrize("use_combined_loader", [False, True])
+@pytest.mark.parametrize("use_combined_loader", [False])
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:32]: This too

File: tests/utilities/test_fetching.py [hunk 9]
```diff
@@ -95,16 +97,11 @@ def test_misconfiguration_error():
 
 def get_cycles_per_ms() -> float:
     """
-    Get 10 values and remove the 2 max and 2 min and return the avg.
-    This is to avoid system disturbance that skew the results, e.g. the very first cuda call likely does a bunch of
-    init, which takes much longer than subsequent calls.
+    Measure and return approximate number of cycles per millisecond for torch.cuda._sleep
+    Copied from: github.com/pytorch/pytorch/blob/master/test/test_cuda.py
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:101]: This too

File: tests/loops/test_iterator_batch_processor.py [hunk 10]
```diff
@@ -73,6 +73,8 @@ def training_step(self, dataloader_iter: Iterator) -> STEP_OUTPUT:
     def train_dataloader(self):
         return DataLoader(RandomDataset(_BATCH_SIZE, _DATASET_LEN))
 
+    training_epoch_end = None
```
Comment (justusschock) [inline tests/loops/test_iterator_batch_processor.py:76]: why is this necessary?

File: tests/utilities/test_fetching.py [hunk 11]
```diff
@@ -114,6 +111,14 @@ def measure() -> float:
         cycles_per_ms = 1000000 / start.elapsed_time(end)
         return cycles_per_ms
 
+    # Get 10 values and remove the 2 max and 2 min and return the avg.
+    # This is to avoid system disturbance that skew the results, e.g.
+    # the very first cuda call likely does a bunch of init, which takes
+    # much longer than subsequent calls.
+    #
+    # Tested on both Tesla V100, Quadro GP100, Titan RTX, RTX 3090 GPUs
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:119]: This too

**Comment by justusschock** [comment]
lol @carmocca  was 10 secs faster than me with most of his comments :D

File: pytorch_lightning/loops/processors/iterator_batch_processor.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/processors/iterator_batch_processor.py:133]: Done !

File: pytorch_lightning/loops/processors/iterator_batch_processor.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/loops/processors/iterator_batch_processor.py]: Done !

File: tests/utilities/test_fetching.py (hunk 8 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:32]: Done !

File: tests/utilities/test_fetching.py (hunk 9 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:101]: Removed !

File: tests/utilities/test_fetching.py (hunk 11 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:119]: removed !

File: pytorch_lightning/loops/fit_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:240]: Removed directly !

File: pytorch_lightning/loops/processors/iterator_batch_processor.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/processors/iterator_batch_processor.py:133]: Done !

File: pytorch_lightning/loops/processors/iterator_batch_processor.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/loops/processors/iterator_batch_processor.py:162]: Done !

File: tests/loops/test_iterator_batch_processor.py (hunk 10 above)
Comment (tchaton) [reply tests/loops/test_iterator_batch_processor.py:76]: Removed !

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:96]: Not yet, but we will think about this when made public.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:87]: Yes, this is why the InterBatchProcessor was introduced. This code is meant to replace it.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 7 above)
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/data_connector.py:96]: Maybe raise a warning when `PL_INTER_BATCH_PARALLELISM` is True for non GPU accelerators

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 12]
```diff
@@ -66,15 +79,38 @@ def on_trainer_init(
...
+            # note: this is an experimental feature
+            return InterBatchParallelDataFetcher()
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/data_connector.py:98]: Should we pass `prefetch_batches` > 1 for better results for the test?

File: tests/utilities/test_fetching.py (hunk 9 above)
Comment (kaushikb11) [reply tests/utilities/test_fetching.py:101]: Why remove it?

Comment (kaushikb11) [reply tests/utilities/test_fetching.py:101]: Don't mind the detailed info

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 13]
```diff
@@ -66,15 +79,38 @@ def on_trainer_init(
...
         else:
-            self.data_fetcher.setup(train_dataloader)
-        prefetcher_iter = iter(self.data_fetcher)
-        profiled_dl = self.trainer.profiler.profile_iterable(enumerate(prefetcher_iter), "get_train_batch")
-        return profiled_dl
+            return DataFetcher()
+
+    def get_profiled_dataloader(self, dataloader: Iterable, dataloader_idx: int = 0) -> Iterable:
+        stage: str = self.trainer.state.stage.value
+        self.data_fetcher = self._select_data_fetcher()
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/data_connector.py:104]:
```suggestion
		if self.data_fetcher is None:
        	self.data_fetcher = self._select_data_fetcher()
```

File: tests/utilities/test_fetching.py [hunk 14]
```diff
@@ -174,15 +175,22 @@ def test_dataloader(self):
...
+        max_epochs=1,
+        gpus=1,
+        limit_train_batches=4,
+        limit_val_batches=0,
+        num_sanity_val_steps=0,
+    )
+
+    with mock.patch.dict(os.environ, {"PL_INTER_BATCH_PARALLELISM": "1"}):
+        t0 = time()
+        trainer = Trainer(**trainer_kwargs)
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:190]: should the trainer creation be timed too?

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:96]: Done !

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:98]: No, there is no need right now :)

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 15]
```diff
@@ -135,11 +135,6 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
             # ------------------------------------
             # TRAINING_STEP + TRAINING_STEP_END
             # ------------------------------------
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:137]: I think this can be simplified. What do you think?

```python
    def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
        """Runs a single training batch.

        Args:
            dataloader_iter: the iterator over the dataloader producing the new batch

        Raises:
            StopIteration: When the epoch is canceled by the user returning -1
        """
        takes_iterator = isinstance(self.batch_loop, IteratorBatchProcessor)
        if takes_iterator:
            batch_input = (dataloader_iter, )
        else:
            _, (batch, is_last) = next(dataloader_iter)
            batch_input = (batch, self.batch_idx)

        self.batch_progress.increment_ready()

        # ------------------------------------
        # TRAINING_STEP + TRAINING_STEP_END
        # ------------------------------------
        with self.trainer.profiler.profile("run_training_batch"):
            batch_output = self.batch_loop.run(*batch_input)

        self.batch_progress.increment_processed()

        # By contract, when taking `dataloader_iter` as an argument, `training_step` is responsible for reporting
        # `is_last` in the result dict, which is used to determine the stop condition for the epoch. So as long as
        # `advance` is invoked, it's correct to assume that there are more batches to be processed.
        self.is_last_batch = is_last if not takes_iterator else batch_output.is_last
```

Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:137]: This would be entirely removed in next PR.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:104]: Doesn't work if the DataFetcher isn't the same for each stage yet.

File: tests/utilities/test_fetching.py (hunk 14 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:190]: I don't think this is an issue as it is counted on both side.

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 16]
```diff
@@ -66,15 +79,40 @@ def on_trainer_init(
         self.trainer.reload_dataloaders_every_n_epochs = reload_dataloaders_every_n_epochs
         self.trainer._is_data_prepared = False
 
-    def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
-        # FIXME: Temporary hack
-        if isinstance(self.data_fetcher, InterBatchParallelDataFetcher):
-            self.data_fetcher.setup(train_dataloader, batch_to_device=self.trainer.accelerator.batch_to_device)
+    def _check_training_step_requires_dataloader_iter(self) -> bool:
+        if not self.trainer.training:
+            return False
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:84]: This if should be placed where this function is called, not inside the function.

`if self.trainer.training and self._check_training_step_requires_dataloader_iter()`

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 17]
```diff
@@ -66,15 +79,40 @@ def on_trainer_init(
...
+        training_step_fx = getattr(self.trainer.lightning_module, "training_step")
+        contains_dataloader_iter = is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True)
+        return contains_dataloader_iter
+
+    def _select_data_fetcher(self) -> AbstractDataFetcher:
+        if self._check_training_step_requires_dataloader_iter():
+            rank_zero_warn(
+                "Found `dataloader_iter` argument in the `training_step`. Note that the support for "
+                "this signature is experimental and the behavior may subject to change."
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:93]: Choose one:

```suggestion
                "this signature is experimental and the behavior is subject to change."
```

```suggestion
                "this signature is experimental and the behavior may change."
```

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 18]
```diff
@@ -66,15 +79,40 @@ def on_trainer_init(
...
+            # note: this is an experimental feature
+            if not self.trainer.training_type_plugin.on_gpu:
+                raise MisconfigurationException("Inter batch parallelism is available only when using Nvidia GPUs.")
+            return InterBatchParallelDataFetcher()
         else:
-            self.data_fetcher.setup(train_dataloader)
-        prefetcher_iter = iter(self.data_fetcher)
-        profiled_dl = self.trainer.profiler.profile_iterable(enumerate(prefetcher_iter), "get_train_batch")
-        return profiled_dl
+            return DataFetcher()
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:102]: nit: unnecessary else

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 19]
```diff
@@ -66,15 +79,40 @@ def on_trainer_init(
...
+
+    def get_profiled_dataloader(self, dataloader: Iterable, dataloader_idx: int = 0) -> Iterable:
+        stage: str = self.trainer.state.stage.value
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/data_connector.py:105]: Is there a reason why we don't pass the Enum?

You could use the `stage.dataloader_prefix` property if you want to match the batch names with the dataloader names

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:84]: Done !

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:93]: Done !

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 18 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:102]: Done !

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 19 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:105]: This is used only for profiling and we didn't have this feature before. Worth to make another PR once finalised.

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 20]
```diff
@@ -91,8 +91,8 @@ def advance(
         if batch is None:
             raise StopIteration
 
-        with self.trainer.profiler.profile("evaluation_batch_to_device"):
-            batch = self.trainer.accelerator.batch_to_device(batch, dataloader_idx=dataloader_idx)
+        if not self.trainer.data_connector.evaluation_data_fetcher.store_on_device:
+            batch = self.trainer.accelerator.batch_to_device(batch)
```
Comment (justusschock) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:95]: What about the prediction loop? Do we need to care for that as well?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 20 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:95]: It is not enable right now, but might come in future PRs :)

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 21]
```diff
@@ -280,6 +280,8 @@ def _training_step(
                 training_step_output = self.trainer.accelerator.training_step(step_kwargs)
                 self.trainer.accelerator.post_training_step()
 
+            del step_kwargs
```
Comment (aazzolini) [inline pytorch_lightning/loops/batch/training_batch_loop.py:283]: Why is this needed?

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 22]
```diff
@@ -91,8 +91,9 @@ def advance(
         if batch is None:
             raise StopIteration
 
-        with self.trainer.profiler.profile("evaluation_batch_to_device"):
-            batch = self.trainer.accelerator.batch_to_device(batch, dataloader_idx=dataloader_idx)
+        if not self.trainer.data_connector.evaluation_data_fetcher.store_on_device:
```
Comment (aazzolini) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:94]: Do we plan to remove this?

File: pytorch_lightning/utilities/fetching.py [hunk 23]
```diff
@@ -47,22 +47,27 @@ class SimpleDataFetcher(AbstractDataFetcher):
             def fetching_function(self):
                 while True:
                     try:
-                        yield next(self.dataloader_iter), False
+                        return next(self.dataloader_iter), False
```
Comment (aazzolini) [inline pytorch_lightning/utilities/fetching.py:50]: Why this change?
I am having a hard time to undertsand why we have a while loop that will only run for one iteration. It's not clear why this hasn't been replaced by an if statement, or why we didn't just remove the while loop here.

Also it seems that this function now became just semantically equivalent to implement __next__ in a regular iterator, why do we still need this fecthing_function here?

**Review by aazzolini — COMMENTED** [review]
A few inline comments while  trying to undestand the reasoning around some of the complexity in this diff and other diffs in the stack.

**Comment by carmocca** [comment]
@aazzolini I also noticed some of the things you mention here and opened https://github.com/PyTorchLightning/pytorch-lightning/pull/9058 to address them. Thanks for reviewing!
